managed-ledger/src/main/java/org/apache/bookkeeper/mledger/LedgerOffloader.java

```diff
@@ -35,6 +39,78 @@
 @InterfaceStability.Evolving
 public interface LedgerOffloader {
 
+    @ToString
+    class SegmentInfo {
+        public SegmentInfo(UUID uuid, long beginLedger, long beginEntry, String driverName,
+                           Map<String, String> driverMetadata) {
+            this.uuid = uuid;
+            this.beginLedger = beginLedger;
+            this.beginEntry = beginEntry;
+            this.driverName = driverName;
+            this.driverMetadata = driverMetadata;
+        }
+
+
+        public final UUID uuid;
+        public final long beginLedger;
+        public final long beginEntry;
+        public final String driverName;
+        volatile private long endLedger;
+        volatile private long endEntry;
+        volatile boolean closed = false;
+        public final Map<String, String> driverMetadata;
+
+        public boolean isClosed() {
+            return closed;
+        }
+
+        public void closeSegment(long endLedger, long endEntry) {
+            this.endLedger = endLedger;
+            this.endEntry = endEntry;
+            this.closed = true;
+        }
+
+        public OffloadResult result() {
+            return new OffloadResult(beginLedger, beginEntry, endLedger, endEntry);
+        }
+    }
+
+
+    class OffloadResult {
+        public final long beginLedger;
+        public final long beginEntry;
+        public final long endLedger;
+        public final long endEntry;
+
+        public OffloadResult(long beginLedger, long beginEntry, long endLedger, long endEntry) {
+            this.beginLedger = beginLedger;
+            this.beginEntry = beginEntry;
+            this.endLedger = endLedger;
+            this.endEntry = endEntry;
+        }
+    }
+
+    /**
+     * Used to store driver info, buffer entries, mark progress, etc.
+     * Create one per second.
+     */
+    interface OffloaderHandle {
+
+        /**
+         * return true when both buffer have enough size and ledger/entry id is next to the current one.
+         * @param size
+         * @return
+         */
+        boolean canOffer(long size);
+
+        PositionImpl lastOffered();
+
+        boolean offerEntry(EntryImpl entry) throws OffloadSegmentClosedException,
```

Comment:
It's better to also add async method for these 3 methods. Usually, we should implement the sync method based on the async method.

Reply:
+1